runtime/domain.c

```diff
@@ -45,20 +45,25 @@
 #include "caml/sync.h"
 #include "caml/weak.h"
 
-/* From a runtime perspective, domains must handle stop-the-world (STW)
-   sections, during which:
-    - they are within a section no mutator code is running
-    - all domains will execute the section in parallel
-    - barriers are provided to know all domains have reached the
-      same stage within a section
+/* The runtime can run stop-the-world (STW) sections, during which all
+   active domains run the same code in parallel (with a barrier
+   mechanism to synchronize within this code). See
+   [caml_try_run_on_all_domains] below.
 
    Stop-the-world sections are used to handle duties such as:
     - minor GC
-    - major GC to trigger major state machine phase changes
-
-   Two invariants for STW sections:
-    - domains only execute mutator code if in the stop-the-world set
-    - domains in the stop-the-world set guarantee to service the sections
+    - major GC phase changes
+
+   We guarantee that no mutator code runs in parallel with a STW
+   section, and neither does domain initialization or cleanup code.
```

Comment:
We actually run cleanup code for termination in parallel with a STW section (e.g. `caml_teardown_major_gc`). The only guarantee is that a domain in the 'STW set' must participate in the STW section. Once a terminating domain has left the STW set, then it has no synchronisation guarantee with respect to STW sections. 
```suggestion
   section, and domains are blocked from entering or leaving the set of active
   STW participants while a STW section is in progress.
```